WalletWasabi/Models/CoinsRegistry.cs

```diff
@@ -0,0 +1,168 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Specialized;
+using System.Linq;
+using NBitcoin;
+using WalletWasabi.Models;
+
+namespace WalletWasabi.Models
+{
+	public class CoinsRegistry : ICoinsView
+	{
+		private HashSet<SmartCoin> Coins { get; }
+		private HashSet<SmartCoin> LatestCoinsSnapshot { get; set; }
+		private bool InvalidateSnapshot { get; set; }
+		private object Lock { get; set; }
+
+		public event NotifyCollectionChangedEventHandler CollectionChanged;
+
+		public CoinsRegistry()
+		{
+			Coins = new HashSet<SmartCoin>();
+			LatestCoinsSnapshot = new HashSet<SmartCoin>();
+			InvalidateSnapshot = false;
+			Lock = new object();
+		}
+
+		private CoinsView AsCoinsView()
+		{
+			lock (Lock)
+			{
+				if (InvalidateSnapshot)
+				{
+					LatestCoinsSnapshot = new HashSet<SmartCoin>(Coins);
+					InvalidateSnapshot = false;
+				}
+			}
+			return new CoinsView(LatestCoinsSnapshot);
+		}
+
+		public bool IsEmpty => !AsCoinsView().Any();
+
+		public SmartCoin GetByOutPoint(OutPoint outpoint)
+		{
+			return AsCoinsView().FirstOrDefault(x => x.GetOutPoint() == outpoint);
+		}
+
+		public bool TryAdd(SmartCoin coin)
+		{
+			var added = false;
+			lock (Lock)
+			{
+				added = Coins.Add(coin);
+				InvalidateSnapshot |= added;
+			}
+
+			if (added)
+			{
+				CollectionChanged?.Invoke(this, new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Add, coin));
+			}
+			return added;
+		}
+
+		public void Remove(SmartCoin coin)
+		{
+			var coinsToRemove = AsCoinsView().DescendantOf(coin).ToList();
+			coinsToRemove.Add(coin);
+			var removedCoins = new List<SmartCoin>();
+			lock (Lock)
+			{
+				foreach (var toRemove in coinsToRemove)
+				{
+					if(Coins.Remove(toRemove))
```

Comment:
```suggestion
					if (Coins.Remove(toRemove))
```
@nopara73 #2392 helps to fix this.